internal/cmd/shell.go

```diff
@@ -90,7 +91,113 @@ func shell(cmd *cobra.Command, cmdArgs []string) error {
 
 	envName := cmdArgs[0]
 
-	if err := runCommand(ctx, envName, command, args); err != nil {
+	// Before the command is run, ensure the workspace is on and ready to accept
+	// an ssh connection.
+	client, err := newClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	env, err := findEnv(ctx, client, envName, coder.Me)
+	if err != nil {
+		return err
+	}
+
+	var requiredRebuilds []string
+
+	// check if there are required rebuilds
+	for _, r := range env.RebuildMessages {
+		if r.Required {
+			requiredRebuilds = append(requiredRebuilds, clog.Causef(r.Text))
+		}
+	}
+
+	switch {
+	// If this conditonal is true, a rebuild is **required** to make
+	// the sh command work.
+	case env.LatestStat.ContainerStatus == coder.EnvironmentOff || len(requiredRebuilds) > 0:
+		// TODO: (@emyrk) I'd like to add a --force and --verbose flags to this command,
+		//					but currently DisableFlagParsing is set to true.
+		//					To enable force/verbose, we'd have to parse the flags ourselves,
+		//					or make the user `coder sh <env> -- [args]`
+		//
+
+		var rebuildReason string // Either because it is off, or a rebuild is required
+		if env.LatestStat.ContainerStatus == coder.EnvironmentOff {
+			rebuildReason = fmt.Sprintf("Environment %q is currently not running.", env.Name)
+		} else {
+			// Print out the reasons the rebuild is required. Newlines don't work well with
+			// the prompt label, so printing it above the prompt will have to do.
+			clog.LogWarn("rebuild required", requiredRebuilds...)
+			rebuildReason = fmt.Sprintf("Environment %q requires a rebuild to work correctly.", env.Name)
+		}
+
+		confirm := promptui.Prompt{
+			Label:     rebuildReason + " Do you wish to rebuild it now? (this will take a moment)",
+			IsConfirm: true,
+		}
+		if _, err := confirm.Run(); err != nil {
+			// User selected not to rebuild :(
+			return clog.Fatal(
+				"environment is not ready for use",
+				"environment requires a rebuild",
+				fmt.Sprintf("its current status is %q", env.LatestStat.ContainerStatus),
+				clog.BlankLine,
+				clog.Tipf("run \"coder envs rebuild %s --follow\" to start the environment", env.Name),
+			)
+		}
+
+		// Start the rebuild
+		if err = client.RebuildEnvironment(ctx, env.ID); err != nil {
+			return err
+		}
+
+		fallthrough // Fallthrough to watching the logs
+	case env.LatestStat.ContainerStatus == coder.EnvironmentCreating:
+		// Environment is in the process of being created, just trail the logs
+		// and wait until it is done
+		clog.LogInfo(fmt.Sprintf("Environment %q is being rebuilt to be ready for use.", env.Name))
+
+		// Watch the rebuild.
+		// TODO: (@emyrk) Does trailBuildLogs ensure the container is "ON" when it returns?
+		//		Or do I need to poll `findEnv` until it is 'ON'
+		if err = trailBuildLogs(ctx, client, env.ID); err != nil {
```

Comment:
you're exactly right, `trailBuildLogs` exiting is not indicative of a _successful_ build.

Reply:
Added a retry/wait mechanism to check for it, then I re-read this.

When the `trailBuildLogs` finishes, is it possible to still be in `CREATING` state? 
I will check if it's in the `OFF` or `FAILED`. I only need the wait mechanism if it might still be in the `CREATING` state, like a small race condition or something.

Reply:
If `trailBuildLogs` completes without an error, I think we can assume it is not `CREATING`.

Reply:
I'll remove my polling then.